File: src/python/problems/DeleteDuplicates.py
#https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class DeleteDuplicates:
    def Solution(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = ListNode(next=head)
        head = prev
        current = head.next
        remove = False
        while current:
            if current.next and current.val == current.next.val:
                prev.next = current.next
                remove = True
            elif remove:
                prev.next = current.next
                remove = False
            else:
                prev = current

            current = current.next

        return head.next


DeleteDuplicates().Solution(ListNode(1, ListNode(2, ListNode(2))))

File: src/csharp/Problems/DeleteDuplicates.cs
//https://leetcode.com/problems/remove-duplicates-from-sorted-list/

namespace LeetCode.Problems;

public sealed class DeleteDuplicates : ProblemBase
{
    [Theory]
    [ClassData(typeof(DeleteDuplicates))]
    public override void Test(object[] data) => base.Test(data);

    protected override void AddTestCases()
        => Add(it => it.ParamListNode("[1,1,2]").ResultListNode("[1,2]"))
          .Add(it => it.ParamListNode("[]").ResultListNode("[]"))
          .Add(it => it.ParamListNode("[1,1,2,3,3]").ResultListNode("[1,2,3]"));

    private ListNode? Solution(ListNode? head)
    {
        var prev = head;
        var current = head?.next;
        while (current != null)
        {
            if (prev!.val == current.val)
            {
                prev.next = current.next;
            }
            else
            {
                prev = current;
            }

            current = current.next;
        }

        return head;
    }
}